### PR TITLE
[Clock] initialize the current time with midnight before modifying the date

### LINKED ISSUE #54497: [Clock] DatePoint created with fixed date should be set at midnight
### Symfony version(s) affected

6.4.5

### Description

PHP: 8.1

> The DatePoint class extends DateTimeImmutable, so you can use it in your code anywhere a DateTimeImmutable or DateTimeInterface is expected.

I would expect that the behavior of DatePoint does not differ from DateTimeImmutable to use as a pure replacement. However, I cannot use it as it is everywhere (including in test cases) because creating a date with a fixed date keep the current time.

See the repro below.

### How to reproduce

```php
<?php

$datePoint = new \Symfony\Component\Clock\DatePoint('2024-01-01');
$immutableDate = new \DateTimeImmutable('2024-01-01');

\PHPUnit\Framework\Assert::assertEquals('2024-01-01 00:00:00', $immutableDate->format('Y-m-d H:i:s')); // Pass
\PHPUnit\Framework\Assert::assertEquals('2024-01-01 00:00:00', $datePoint->format('Y-m-d H:i:s')); // Fails
```

### Possible Solution

* Add documentation about the different behavior and that DatePoint differs from DateTimeImmutable (not a pure replacement)
* Or introduce a named constructor keep the different behavior and leave the native constructor 

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #54497
| License       | MIT

### CODE REVIEW

**Review by derrabus — APPROVED** [review]

**Comment by derrabus** [comment]
Mind fabbot. 🙂

**Comment by nicolas-grekas** [comment]
I think we can do something more robust thanks to [date_parse](https://php.net/date-parse), which will tell us when relative dates are used. I'm giving it a try.

File: src/Symfony/Component/Clock/DatePoint.php [hunk 1]
```diff
@@ -32,15 +32,21 @@ public function __construct(string $datetime = 'now', ?\DateTimeZone $timezone =
...
                 }
             } else {
-                $timezone = (new parent($datetime, $timezone ?? $now->getTimezone()))->getTimezone();
+                $builtInDate = new parent($datetime, $timezone ?? $now->getTimezone());
+                $timezone = $builtInDate->getTimezone();
             }
 
             $now = $now->setTimezone($timezone)->modify($datetime);
+
+            if ('00:00:00' === $builtInDate->format('H:i:s')) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Clock/DatePoint.php]:
```suggestion
            if ('00:00:00.000000' === $builtInDate->format('H:i:s.u')) {
```

File: src/Symfony/Component/Clock/Tests/DatePointTest.php [hunk 2]
```diff
@@ -57,4 +57,11 @@ public function testModify()
         $this->expectExceptionMessage('Failed to parse time string (Bad Date)');
         $date->modify('Bad Date');
     }
+
+    public function testTimeDefaultsToMidnight()
+    {
+        $date = new DatePoint('2024-04-09');
+
+        $this->assertSame('2024-04-09 00:00:00', $date->format('Y-m-d H:i:s'));
+    }
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Clock/Tests/DatePointTest.php]:
```suggestion
    /**
     * @testWith ["2024-04-01 00:00:00.000000", "2024-04"]
     *           ["2024-04-09 00:00:00.000000", "2024-04-09"]
     *           ["2024-04-09 03:00:00.000000", "2024-04-09 03:00"]
     *           ["2024-04-09 00:00:00.123456", "2024-04-09 00:00:00.123456"]
     */
    public function testTimeDefaults(string $expected, string $datetime)
    {
        $date = new \DateTimeImmutable($datetime);
        $this->assertSame($expected, $date->format('Y-m-d H:i:s.u'));

        $date = new DatePoint($datetime);
        $this->assertSame($expected, $date->format('Y-m-d H:i:s.u'));
    }
```

**Review by nicolas-grekas — COMMENTED** [review]
I spent too many hours on this, thanks for the patch, I only have minor improvements to suggest!

**Comment by xabbuh** [comment]
@nicolas-grekas done

**Review by nicolas-grekas — APPROVED** [review]

**Pull request merged**